### PR TITLE
[OSD-19745] - Make OCM Agent Highly Available

### PR DESCRIPTION
**Pull request by xiaoyu74**
### What type of PR is this?
_(feature)_

### What this PR does / why we need it?
Update the OCM agent deployment configuration to tolerate disruptions, such as infra nodes being terminated to make OCM Agent Highly Available

### Which Jira/Github issue(s) this PR fixes?
https://issues.redhat.com/browse/OSD-19745

### Test Steps in a staging cluster:
https://gitlab.cee.redhat.com/-/snippets/7462

### CODE REVIEW

File: pkg/ocmagenthandler/ocmagenthandler_deployment.go
Comment (bmeng) [inline pkg/ocmagenthandler/ocmagenthandler_deployment.go:159]: Do we need a separated function for this? And there is no parameter needed for this function

Comment (xiaoyu74) [reply pkg/ocmagenthandler/ocmagenthandler_deployment.go:159]: Yeah, I was thinking to follow the similar pattern as namespace.go in util to decouple this anti-pod-affinity function. It might be more readable to just put the anti-pod-affinity setting here directly.

File: pkg/ocmagenthandler/ocmagenthandler_deployment.go
Comment (xiaoyu74) [reply pkg/ocmagenthandler/ocmagenthandler_deployment.go:253]: Yeah, that makes sense, will do, thanks Bo.

**Review by hectorakemp — COMMENTED** [review]
Pretty much LGTM if tested!

**Comment by xiaoyu74** [comment]
> Pretty much LGTM if tested!

Local test succeed: https://gitlab.cee.redhat.com/-/snippets/7462

File: deploy/20_ocm-agent-operator.Role.yaml
Comment (bmeng) [inline deploy/20_ocm-agent-operator.Role.yaml:7]: Does the intent matter? It is different with before

File: pkg/ocmagenthandler/ocmagenthandler_pdb.go
Comment (bmeng) [inline pkg/ocmagenthandler/ocmagenthandler_pdb.go:28]: That means we will have to have at least 2 replicas, or else it will be non-disruptable 
Wondering if it would be better that we set this based on the ocmagent value? Not sure if it is needed.

File: deploy/20_ocm-agent-operator.Role.yaml
Comment (xiaoyu74) [reply deploy/20_ocm-agent-operator.Role.yaml:7]: I am not sure if the previous format without indentation for this `20_ocm-agent-operator.Role.yaml` has any impact but I just make 2 space indentation in `rules` section in this file to align with the pattern of other RBAC file such as `20_ocm-agent-operator.ClusterRole.yaml` and `20_ocm-agent.Role.yaml` and it works all good.

File: pkg/ocmagenthandler/ocmagenthandler_pdb.go
Comment (xiaoyu74) [reply pkg/ocmagenthandler/ocmagenthandler_pdb.go:28]: Thanks for your review, not sure if I understood your comment correctly.  I have create a PR to set the ocm-agent replicas to 2 in the [Operator Lifecycle Manager (OLM) template](https://github.com/openshift/ocm-agent/blob/e224d04693f988e5f3758d2c10ab3a137e6d6a2d/hack/00-ocm-agent.selectorsyncset.yaml.tmpl). Once it's merged, it can ensure that we have redundancy in our deployment and can provide HA.

With the `MinAvailable` set to 1 in the PDB, we ensure that at least one of these replicas is always available. This is aligned with our goal of maintaining high availability while allowing room for disruptions like maintenance or updates? 

Please correct me if I was wrong.

Comment (bmeng) [reply pkg/ocmagenthandler/ocmagenthandler_pdb.go:28]: yeah, I think it should be good.

**Review by hectorakemp — APPROVED** [review]

**Pull request merged**
